### DISCUSSION TITLE
MacOS

### DISCUSSION THREAD

**Discussion by helge17** [helge17/tuxguitar#250] [General]
Currently I am building TuxGuitar for MacOS on an older MacOS 11 x86_64 VM.

I wanted to migrate to a cleaner MacOS 14 x86_64 VM with Xcode installed. Unfortunately I have a problem with SWT on MacOS 14 and some other MacOS specific questions, so I'm starting this discussion here. Maybe someone can help.

### SWT

~~While the SWT builds from MacOS 11 work fine on 14, when I build the same code on MacOS 14, some GUI elements are invisible when starting TuxGuitar:~~
Update: I can now build and run TuxGuitar on macOS 14 after upgrading SWT to 4.33 (commit ad5c896980c585dec33e12454a35ae77db2a8950).

![MacOS14_SWT](https://github.com/helge17/tuxguitar/assets/5897631/e0e56fc5-075c-4cf0-94d2-94cf62238451)

~~You can click where you would expect the missing elements to be and they react, but they are simply invisible.~~

~~Upgrading SWT to version 4.21 or 4.26 did not fix the issue.~~

~~This bug currently prevents me from migrating to MacOS 14 and mybe care about the next issues:~~

#### SWT - application menu

The application menu items that MacOS adds to TuxGuitar are not translated.

![MacOS14_SWT_app_menu](https://github.com/helge17/tuxguitar/assets/5897631/46b98474-0c78-41df-af95-5429a1331e9f)

#### SWT - Cocoa integration plugin

~~Same question: What is this plugin good for? I can't tell the difference whether the plugin is activated or not.~~
Update: Since PR #335, tg files can be opened with a double click in Finder if this plugin is activated. According to the source code, there should be other advantages, but I haven't noticed them yet.

~~The plugin causes errors on newer SWT versions (SWT 4.14 is the last working version, 4.15 and above show the error):~~
Update: Fixed in #355. TuxGuitar can now be built with SWT 4.26 and runs fine.

![MacOS14_SWT_cocoa-plugin](https://github.com/helge17/tuxguitar/assets/5897631/3b4aa009-8bdd-4473-a3d0-5f63109d6fa4)

```
MacOS14 00-Binary_Packages % ./tuxguitar-2024-02-22-snapshot-macosx-swt-cocoa-x86_64.app/Contents/MacOS/tuxguitar.sh
org.herac.tuxguitar.util.plugin.TGPluginException: org.eclipse.swt.internal.cocoa.NSWindow.standardWindowButton(long)
	at org.herac.tuxguitar.cocoa.toolbar.MacToolbarPlugin.connect(MacToolbarPlugin.java:28)
	at org.herac.tuxguitar.util.plugin.TGPluginManager.connectPlugin(TGPluginManager.java:88)
	at org.herac.tuxguitar.util.plugin.TGPluginManager.connectEnabled(TGPluginManager.java:65)
	at org.herac.tuxguitar.app.TuxGuitar.startUIContext(TuxGuitar.java:141)
	at org.herac.tuxguitar.app.TuxGuitar.access$100(TuxGuitar.java:69)
	at org.herac.tuxguitar.app.TuxGuitar$1$1.run(TuxGuitar.java:124)
	at org.herac.tuxguitar.app.synchronizer.TGSynchronizerControllerImpl$1$1.run(TGSynchronizerControllerImpl.java:29)
	at org.eclipse.swt.widgets.RunnableLock.run(RunnableLock.java:40)
	at org.eclipse.swt.widgets.Synchronizer.runAsyncMessages(Synchronizer.java:185)
	at org.eclipse.swt.widgets.Display.runAsyncMessages(Display.java:4262)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3885)
	at org.herac.tuxguitar.ui.swt.SWTApplication.start(SWTApplication.java:64)
	at org.herac.tuxguitar.app.TuxGuitar.createUIContext(TuxGuitar.java:117)
	at org.herac.tuxguitar.app.TuxGuitar.createApplication(TuxGuitar.java:102)
	at org.herac.tuxguitar.app.TGMainSingleton.launchTuxGuitar(TGMainSingleton.java:57)
	at org.herac.tuxguitar.app.TGMainSingleton.launchSingleton(TGMainSingleton.java:45)
	at org.herac.tuxguitar.app.TGMainSingleton.main(TGMainSingleton.java:35)
Caused by: java.lang.NoSuchMethodException: org.eclipse.swt.internal.cocoa.NSWindow.standardWindowButton(long)
	at java.base/java.lang.Class.getMethod(Class.java:2395)
	at org.herac.tuxguitar.cocoa.TGCocoa.invokeMethod(TGCocoa.java:118)
	at org.herac.tuxguitar.cocoa.TGCocoa.getStandardWindowButton(TGCocoa.java:66)
	at org.herac.tuxguitar.cocoa.toolbar.MacToolbar.init(MacToolbar.java:59)
	at org.herac.tuxguitar.cocoa.toolbar.MacToolbarPlugin.connect(MacToolbarPlugin.java:25)
	... 16 more
```

### JFX

The JFX build on MacOS 14 works as good as on 11 after upgrading JFX to version 21.0.2 with commit 3289a8bb0f20731faea05487614f832f00033b07.

With JFX 17.0.2 TuxGuitar crashed on starup:

```
2024-02-17 20:07:26.630 java[1141:10867] *** Assertion failure in -[_NSTrackingAreaAKViewHelper removeTrackingRect:], _NSTrackingAreaAKManager.m:1780
*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: '0x0 is an invalid NSTrackingRectTag. Common possible reasons for this are: 1. already removed this trackingRectTag, 2. Truncated the NSTrackingRectTag to 32bit at some point.'
```

As suggested in https://github.com/Col-E/Recaf/issues/636#issuecomment-1913914385, upgrading JFX helped.

~~But unlike other MacOS apps, TuxGuitar's menu is not integrated into the MacOS menu bar on top of the screen, but shown in the TuxGuitar window (mentioned in #213; although I cannot reproduce the other problems described in the ticket).~~
Update: TuxGuitar's menu is now integrated into the MacOS menu bar with commit adccd4911891cefc7d8582519708a8c17c67c99f.

![MacOS14_JFX](https://github.com/helge17/tuxguitar/assets/5897631/11dea749-1e8f-439f-8ac2-76856b36dcde)

### CoreAudio and AudioUnit plugin

The CoreAudio plugin has already been deprecated (in folder deprecated/TuxGuitar-CoreAudio/), the AudioUnit plugin is still part of TuxGuitar (desktop/TuxGuitar-AudioUnit/).

Both plugins are ment "for using the native Quicktime synthesizer in TuxGuitar on mac OS X.", according to their ~~"How to Install.txt"~~ README files.

So my first question would be: Is it still worth taking care of the plugins or should we also deprecate the AudioUnit plugin?

@Sy256, since you opened issue #77 about the AudioUnit plugin, maybe you can explain what the plugin is good for these days.

~~In case we want to keep the plugins, the first step would be to rebuild the libraries libtuxguitar-coreaudio-jni.jnilib (currently a universal binary for ppc/i386) and libtuxguitar-audiounit-jni.jnilib (universal binary for ppc/i386/x86_64) from source (as universal binaries for x86_64/arm64). Commit a8ba51be4fe8adbedf793af99f01a75906a50da3 re-adds the Makefiles and fixes a few obvious things, but there ist still a lot to do.~~
Commit 604fee7da79bac1aebd1005e5c650d0a343f0480 now allows rebuilding both libs as universal binaries for x86_64/arm64.

### ARM Macs

Our tar.gz packages currently only contain Java binaries for x86_64. Therefore, TuxGuitar on ARM Macs has to run through Rosetta, which can cause additional problems. Sooner or later we will have to ship a native ARM package (and maybe later switch to pkg format).

### More issues

#116

~~#99~~ #336

**Comment by G-eos** [comment]
Hello,
I build following https://github.com/helge17/tuxguitar/blob/master/INSTALL.md (Mac section) 
- iMac intel, MACOS 13.6.1
- git label bb7ec80a7db78ddc71a028f75b41448338744c61 
- "flavor" tuxguitar-macosx-swt-cocoa-x86_64
The tuxguitar/desktop/build-scripts/tuxguitar-macosx-swt-cocoa-x86_64/target/tuxguitar-SNAPSHOT-macosx-swt-cocoa-x86_64.app does not include jre. I add the jre directory of the 1.6.1 release and it starts smoothly.
Menu appears and about box works.
Note: SWT is the one defined in install.md $-> wget https://archive.eclipse.org/eclipse/downloads/drops4/R-4.13-201909161045/swt-4.13-cocoa-macosx-x86_64.zip

Note : history is working fine to reopen a file.
Note2: launching through command line the tuxguitar.sh has the effect to not have a working menu, unless giving the focus to another application and then come back ... double click on the tuxguitar app from finder does not have this issue.
Note 3: launching with tuxguitar.sh from a shell display the splash screen, which is not the case when double-clicking ...

What can help can I bring there ? I will have a look at #99 but as a newbie on java and tuxguitar and Mac GUI .. it can take some times :)

**Reply by helge17** [reply]
> Note2: launching through command line the tuxguitar.sh has the effect to not have a working menu, unless giving the focus to another application and then come back ... double click on the tuxguitar app from finder does not have this issue.

Mac OS 11: Same issue
Mac OS 14: Menu works immediately, no issue

> Note 3: launching with tuxguitar.sh from a shell display the splash screen, which is not the case when double-clicking ...

Same issue (Mac OS 11 & 14)

**Comment by helge17** [comment]
With the help of [this commit](https://github.com/helge17/tuxguitar-2.0beta/commit/bada12f8) from tuxguitar-2.0beta I was able to rebuild the jni lib for the AudioUnit plugin (and for the deprecated CoreAudio plugin) under MacOS, see commit 604fee7da79bac1aebd1005e5c650d0a343f0480. The lib is now a universal binary for x86_64/arm64. Maybe this fixes problem #77.

The lib is still in the source tree as a binary, but as soon as I switch to MacOS 14 with my build system, the lib can always be built from source.

**Comment by G-eos** [comment]
Hello, I add a GitHub action to build Mac M1.
Please gave a look to this branch. https://github.com/G-eos/tuxguitar/tree/workflow-mac-arm

note, it needs a swt version with arm support. I choose the same one as in your tuxguitar-next branch.
